### PR TITLE
Add support for torch exported model

### PR DESCRIPTION
**Pull request by agunapal**
## Description

This PR adds support for 
- Loading torch.export model
- Includes an example of ResNet 18
- Adds a pytest for torch.export

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

```
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.8.18, pytest-7.3.1, pluggy-1.0.0 -- /home/ubuntu/anaconda3/envs/torchserve/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/ubuntu/serve/test/pytest/.hypothesis/examples')
rootdir: /home/ubuntu/serve
plugins: mock-3.10.0, anyio-3.6.1, cov-4.1.0, hypothesis-6.54.3
collected 1 item                                                                                                                                                                         

test_torch_export.py::test_execute_script_in_custom_directory PASSED                                                                                                               [100%]

==================================================================================== warnings summary ====================================================================================
test_torch_export.py:3
  /home/ubuntu/serve/test/pytest/test_torch_export.py:3: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging

../../../anaconda3/envs/torchserve/lib/python3.8/site-packages/pkg_resources/__init__.py:2871
  /home/ubuntu/anaconda3/envs/torchserve/lib/python3.8/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('mpl_toolkits')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/torchserve/lib/python3.8/site-packages/pkg_resources/__init__.py:2871
../../../anaconda3/envs/torchserve/lib/python3.8/site-packages/pkg_resources/__init__.py:2871
  /home/ubuntu/anaconda3/envs/torchserve/lib/python3.8/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('ruamel')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:242
  /home/ubuntu/anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:242: DeprecationWarning: BILINEAR is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.BILINEAR instead.
    interpolation: int = Image.BILINEAR,

../../../anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:288
  /home/ubuntu/anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:288: DeprecationWarning: NEAREST is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.NEAREST or Dither.NONE instead.
    interpolation: int = Image.NEAREST,

../../../anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:304
  /home/ubuntu/anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:304: DeprecationWarning: NEAREST is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.NEAREST or Dither.NONE instead.
    interpolation: int = Image.NEAREST,

../../../anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:321
  /home/ubuntu/anaconda3/envs/torchserve/lib/python3.8/site-packages/torchvision/transforms/_functional_pil.py:321: DeprecationWarning: BICUBIC is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.BICUBIC instead.
    interpolation: int = Image.BICUBIC,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================= 1 passed, 8 warnings in 8.73s ============================================================================
```

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: examples/pt2/torch_export/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# TorchServe inference with torch export model
```
Comment (msaroufim) [inline examples/pt2/torch_export/README.md:1]: there were some export related details on the main README as well could you consolidate?

File: examples/pt2/torch_export/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+You can find more details [here](https://pytorch.org/docs/stable/export.html)
+
+
+
+### Pre-requisites
+
+- `PyTorch >= 2.1.0`
+
+Change directory to the root of `serve`
```
Comment (msaroufim) [inline examples/pt2/torch_export/README.md:21]: cd serve is simpler

File: examples/pt2/torch_export/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# TorchServe inference with torch export model
+
+This example shows how to run TorchServe a Torch exported model
+
+`torch.export.export()` produces the traced graph representing only the Tensor computation of the function in an Ahead-of-Time fashion. This can then be serialized.
+
+`torch.export.export()` takes a `torch.nn.Module` or a method along with sample inputs, and captures the computation graph into an `torch.export.ExportedProgram`
+
+`torch.export` differs from `torch.compile` is a few ways
+- JIT vs AOT
```
Comment (msaroufim) [inline examples/pt2/torch_export/README.md:10]: this needs a longer elaboration

File: examples/pt2/torch_export/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+The model is saved with `.pt2` extension
+
+```
+python examples/pt2/torch_export/resnet18_torch_export.py
+```
+
+### Create model archive
+
+```
+torch-model-archiver --model-name res18-pt2 --handler image_classifier --version 1.0 --serialized-file resnet18.pt2 --extra-files ./examples/image_classifier/index_to_name.json
```
Comment (msaroufim) [inline examples/pt2/torch_export/README.md:36]: you need to update the archiver code docstrings as well

**Review by msaroufim — CHANGES_REQUESTED** [review]
Right now there is no reason to just export a model, you want to combine this with an aot inductor tutorial so people can get ahead of time compilation and avoid cold starts

File: examples/pt2/torch_export/README.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
+# TorchServe inference with torch export model
+
+This example shows how to run TorchServe a Torch exported model
+
+`torch.export.export()` produces the traced graph representing only the Tensor computation of the function in an Ahead-of-Time fashion. This can then be serialized.
+
+`torch.export.export()` takes a `torch.nn.Module` or a method along with sample inputs, and captures the computation graph into an `torch.export.ExportedProgram`
+
+`torch.export` differs from `torch.compile` is a few ways
```
Comment (msaroufim) [inline examples/pt2/torch_export/README.md:9]: you need to combo aot inductor with this story

**Review by msaroufim — CHANGES_REQUESTED** [review]

**Comment by agunapal** [comment]
torch.export.export by itself only creates an FX Graph representation of the model and serializes it. So, at inference, the model is executed in eager. Hence, even though this works, closing this PR as it doesn't make use of torch.compile

**Pull request closed without merging**